rule/sexpr/lexer.go

```diff
@@ -64,3 +73,385 @@ func isComment(r rune) bool {
+			}
+			return STRING, b.String(), err
+		}
+		if escape {
+			b.WriteRune(rn)
+			escape = false
+			continue
+		}
+		if isString(rn) {
+			break
+		}
+		escape = rn == '\\'
+		if !escape {
+			b.WriteRune(rn)
+		}
+	}
+	return STRING, b.String(), nil
+}
+
+// scanNumber scans a contiguous string representing a number.  As we
+// have to handle the negative numeric form, it's possible that the
+// '-' rune can prefix a number.  This is problematic because '-' can
+// also be a symbol referring to the arithmetic operation "minus" -
+// that confusion is resolved by scanNumber, and should it consider
+// the latter case to be true it will return a SYMBOL rather than a
+// NUMBER.
+func (s *Scanner) scanNumber() (Token, string, error) {
+	var b bytes.Buffer
+
+	// We can be certain this isn't EOF because we will already
+	// have read and unread the rune before arriving here.
+	rn, err := s.readRune()
+	if err != nil {
+		// Something drastic happened, because we read this fine the first time.
+		return NUMBER, "", err
+	}
+
+	// Whatever happens we'll want the rune.
+	b.WriteRune(rn)
+
+	// Deal with the first rune.  Numbers have special rules about
+	// the first rune, specifically it, and only it, may be the
+	// minus symbol. When we loop later any occurrence of '-' will
+	// be an error.
+	if rn == '-' {
+
+		// Now we look ahead to see if a number is coming, if
+		// its anything else then this isn't a negative
+		// number, but some other form.
+		rn, err := s.readRune()
+
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// In reality having '-' as the final
+				// symbol in a stream is never useful,
+				// but this is the sort of error we
+				// should catch in the Parser, not the
+				// scanner.
+				return SYMBOL, b.String(), nil
+			}
+			return NUMBER, b.String(), err
+		}
+
+		// We've stored the rune, and we know we'll want to
+		// unread whatever happens, so lets just do that now.
+		s.unreadRune(rn)
+
+		// If the next rune isn't a digit then we're going to
+		// assume this is the minus operator and return '-' as
+		// a symbol instead of a number.  There are still
+		// cases where this wouldn't be valid, but they're all
+		// errors and we'll leave that for the Parser to
+		// handle.
+		if !unicode.IsDigit(rn) {
+			return SYMBOL, b.String(), nil
+		}
+	}
+
+	// OK, let's scan the rest of the number...
+
+	for {
+		rn, err := s.readRune()
+		if err != nil {
+			se := err.(*ScanError)
+			if se.EOF {
+				// EOF is a valid terminator for a number
+				return NUMBER, b.String(), nil
+			}
+			return NUMBER, b.String(), err
+		}
+		if rn == '-' {
+			// As we said before '-' can't appear in the
+			// body of a number, this is an error.
+			return NUMBER, b.String(), s.newScanError("invalid number format (minus can only appear at the beginning of a number)")
+		}
+
+		// Valid number parts are written to the buffer
+		if isNumber(rn) || rn == '.' {
```

Comment:
If I understand correctly, this `.3..14...1....5.....9.....` won't trigger an error?

Reply:
@asdine you do understand it correctly - we'll validate that it really is a valid number when we do type checking in the parser.